codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -495,20 +518,22 @@ private void incrementPendingBytes(int numBytes) {
              * Note: this does not flush the {@link ChannelHandlerContext}.
              */
             int write(int allowedBytes) {
-                int before = payload.size();
-                needFlush |= payload.write(Math.max(0, allowedBytes));
-                int writtenBytes = before - payload.size();
                 try {
+                    pendingWriteQueue.remove();
```

Comment:
@buchgr 

> If this event needs to be distributed to different areas of the code the why not do that? Set up a listener interface and have encoder implement it and the flow controller will notify this listener after it has done the queue cleanup.

I assume you're referring to the `writeComplete` callback from the flow controller to the encoder? I'm fine with doing that but we still have the problem that the flow controller doesn't control what the encoder does in its `write` callback.  We may change OUR encoder so that it only modifies stream state during `writeComplete` but the flow controller shouldn't trust that a call to `write` with a different encoder will behave the same way.  The flow controller must protect its queue state in all cases.